File: quant/portfolio/rebalancer.py
"""Rebalancing engine: target portfolio → trade list.

Converts a target weight vector into a list of trades needed to move from
the current portfolio to the target, respecting turnover constraints and
producing executable orders compatible with the OMS.
"""
from __future__ import annotations

from dataclasses import dataclass

from quant.portfolio.constraints import PortfolioConstraints


@dataclass(frozen=True, slots=True)
class Trade:
    """A single trade required to move towards the target portfolio.

    Attributes:
        symbol:         Asset identifier.
        current_weight: Current portfolio weight.
        target_weight:  Desired portfolio weight after rebalance.
        trade_weight:   Weight delta to execute (positive = buy, negative = sell).
        dollar_amount:  Approximate dollar value of the trade.
        side:           "BUY" or "SELL".
    """

    symbol: str
    current_weight: float
    target_weight: float
    trade_weight: float
    dollar_amount: float
    side: str


@dataclass(frozen=True, slots=True)
class RebalanceResult:
    """Output of a rebalance computation.

    Attributes:
        trades:            List of trades to execute.
        target_weights:    Final target weights (post-constraint adjustment).
        turnover:          Total one-way turnover (sum of |trade_weight|).
        n_buys:            Number of buy trades.
        n_sells:           Number of sell trades.
        turnover_capped:   True if turnover was reduced to meet constraint.
    """

    trades: list[Trade]
    target_weights: dict[str, float]
    turnover: float
    n_buys: int
    n_sells: int
    turnover_capped: bool


class RebalanceEngine:
    """Converts target portfolio weights into an executable trade list.

    Handles:
    - Dead-band filtering (skip trades below a minimum threshold)
    - Turnover budgeting (proportionally scale trades if over budget)
    - Dollar amount estimation for OMS submission

    Usage::

        engine = RebalanceEngine(min_trade_weight=0.005)
        result = engine.rebalance(
            current_weights={"AAPL": 0.3, "GOOG": 0.2},
            target_weights={"AAPL": 0.25, "GOOG": 0.25, "MSFT": 0.1},
            portfolio_value=1_000_000,
            constraints=PortfolioConstraints(max_turnover=0.3),
        )
        for trade in result.trades:
            print(f"{trade.side} {trade.symbol}: ${trade.dollar_amount:,.0f}")

    Args:
        min_trade_weight: Minimum absolute trade weight to execute (dead band).
            Trades smaller than this are dropped to avoid excessive churn.
    """

    def __init__(self, min_trade_weight: float = 0.001) -> None:
        self._min_trade_weight = min_trade_weight

    def rebalance(
        self,
        current_weights: dict[str, float],
        target_weights: dict[str, float],
        portfolio_value: float,
        constraints: PortfolioConstraints | None = None,
    ) -> RebalanceResult:
        """Compute trades to move from current to target weights.

        Args:
            current_weights:  {symbol: weight} of current portfolio.
            target_weights:   {symbol: weight} of desired portfolio.
            portfolio_value:  Total portfolio value in dollars.
            constraints:      Optional constraints (primarily for turnover budget).

        Returns:
            RebalanceResult with trade list and summary statistics.
        """
        all_symbols = sorted(set(current_weights) | set(target_weights))

        # Compute raw deltas
        deltas: dict[str, float] = {}
        for sym in all_symbols:
            curr = current_weights.get(sym, 0.0)
            tgt = target_weights.get(sym, 0.0)
            delta = tgt - curr
            deltas[sym] = delta

        raw_turnover = sum(abs(d) for d in deltas.values())

        # Apply turnover budget if constrained
        turnover_capped = False
        if constraints and constraints.max_turnover is not None and raw_turnover > constraints.max_turnover + 1e-9:
            scale = constraints.max_turnover / raw_turnover
            deltas = {sym: d * scale for sym, d in deltas.items()}
            turnover_capped = True

        # Build trade list, filtering by dead band
        trades: list[Trade] = []
        final_weights: dict[str, float] = {}

        for sym in all_symbols:
            delta = deltas[sym]
            curr = current_weights.get(sym, 0.0)
            actual_target = curr + delta

            final_weights[sym] = actual_target

            if abs(delta) < self._min_trade_weight:
                continue

            trades.append(
                Trade(
                    symbol=sym,
                    current_weight=curr,
                    target_weight=actual_target,
                    trade_weight=delta,
                    dollar_amount=abs(delta) * portfolio_value,
                    side="BUY" if delta > 0 else "SELL",
                )
            )

        # Remove zero-weight positions from final weights
        final_weights = {s: w for s, w in final_weights.items() if abs(w) > 1e-9}

        actual_turnover = sum(abs(t.trade_weight) for t in trades)
        n_buys = sum(1 for t in trades if t.side == "BUY")
        n_sells = sum(1 for t in trades if t.side == "SELL")

        return RebalanceResult(
            trades=trades,
            target_weights=final_weights,
            turnover=actual_turnover,
            n_buys=n_buys,
            n_sells=n_sells,
            turnover_capped=turnover_capped,
        )

File: quant-rs/quant-portfolio/src/rebalancer.rs
//! Threshold-based portfolio rebalancing with trade generation.
//!
//! Mirrors `quant.portfolio.rebalancer`.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};

// ── Trade ─────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Trade {
    pub symbol: String,
    /// "BUY" or "SELL".
    pub side: String,
    /// Target portfolio weight (post-trade).
    pub target_weight: f64,
    /// Change in weight (can be negative for sells).
    pub trade_weight: f64,
    /// Absolute dollar amount of the trade.
    pub dollar_amount: f64,
}

// ── RebalanceResult ───────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RebalanceResult {
    pub trades: Vec<Trade>,
    /// Total portfolio turnover (sum of absolute weight changes / 2).
    pub turnover: f64,
    pub rebalance_triggered: bool,
}

// ── Rebalancer ────────────────────────────────────────────────────────────────

pub struct Rebalancer {
    /// Minimum total turnover to trigger a rebalance.
    pub threshold: f64,
    /// Dead band for individual trades (trades smaller than this are skipped).
    pub min_trade_weight: f64,
}

impl Default for Rebalancer {
    fn default() -> Self {
        Self {
            threshold: 0.01,
            min_trade_weight: 0.001,
        }
    }
}

impl Rebalancer {
    /// Generate trades to move `current_weights` toward `target_weights`.
    ///
    /// # Arguments
    /// * `target_weights`  — {symbol: weight} from the optimizer.
    /// * `current_weights` — {symbol: weight} from the OMS positions.
    /// * `portfolio_value` — total portfolio value in dollars.
    pub fn rebalance(
        &self,
        target_weights: &HashMap<String, f64>,
        current_weights: &HashMap<String, f64>,
        portfolio_value: f64,
    ) -> RebalanceResult {
        // Compute raw weight deltas (target - current) for each symbol.
        let mut all_symbols: Vec<String> = target_weights
            .keys()
            .chain(current_weights.keys())
            .cloned()
            .collect::<std::collections::HashSet<_>>()
            .into_iter()
            .collect();
        all_symbols.sort();

        let deltas: Vec<(String, f64)> = all_symbols
            .iter()
            .map(|sym| {
                let target = target_weights.get(sym).copied().unwrap_or(0.0);
                let current = current_weights.get(sym).copied().unwrap_or(0.0);
                (sym.clone(), target - current)
            })
            .collect();

        // Total turnover = half of sum of absolute deltas.
        let total_turnover: f64 = deltas.iter().map(|(_, d)| d.abs()).sum::<f64>() / 2.0;

        // Check threshold.
        if total_turnover < self.threshold {
            return RebalanceResult {
                trades: vec![],
                turnover: total_turnover,
                rebalance_triggered: false,
            };
        }

        // Build trades, skipping dead-band.
        let trades: Vec<Trade> = deltas
            .into_iter()
            .filter(|(_, delta)| delta.abs() >= self.min_trade_weight)
            .map(|(symbol, delta)| {
                let side = if delta > 0.0 { "BUY" } else { "SELL" }.to_string();
                let target = target_weights.get(&symbol).copied().unwrap_or(0.0);
                Trade {
                    symbol,
                    side,
                    target_weight: target,
                    trade_weight: delta,
                    dollar_amount: delta.abs() * portfolio_value,
                }
            })
            .collect();

        RebalanceResult {
            trades,
            turnover: total_turnover,
            rebalance_triggered: true,
        }
    }
}

// ── Unit tests ────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_no_rebalance_below_threshold() {
        let rebalancer = Rebalancer {
            threshold: 0.05,
            min_trade_weight: 0.001,
        };
        let target: HashMap<String, f64> = [("AAPL".to_string(), 0.51)].into_iter().collect();
        let current: HashMap<String, f64> = [("AAPL".to_string(), 0.50)].into_iter().collect();
        let result = rebalancer.rebalance(&target, &current, 100_000.0);
        assert!(!result.rebalance_triggered);
        assert!(result.trades.is_empty());
    }

    #[test]
    fn test_rebalance_above_threshold() {
        let rebalancer = Rebalancer::default();
        let target: HashMap<String, f64> = [("AAPL".to_string(), 0.6), ("GOOG".to_string(), 0.4)]
            .into_iter()
            .collect();
        let current: HashMap<String, f64> = [("AAPL".to_string(), 0.3), ("GOOG".to_string(), 0.7)]
            .into_iter()
            .collect();
        let result = rebalancer.rebalance(&target, &current, 100_000.0);
        assert!(result.rebalance_triggered);
        assert_eq!(result.trades.len(), 2);
    }

    #[test]
    fn test_turnover_calculation() {
        let rebalancer = Rebalancer::default();
        let target: HashMap<String, f64> = [("A".to_string(), 1.0)].into_iter().collect();
        let current: HashMap<String, f64> = [("A".to_string(), 0.5), ("B".to_string(), 0.5)]
            .into_iter()
            .collect();
        let result = rebalancer.rebalance(&target, &current, 100_000.0);
        // |1.0-0.5| + |0.0-0.5| = 1.0 → turnover = 0.5
        assert!((result.turnover - 0.5).abs() < 1e-9);
    }

    #[test]
    fn test_buy_sell_classification() {
        let rebalancer = Rebalancer {
            threshold: 0.0,
            min_trade_weight: 0.0,
        };
        let target: HashMap<String, f64> =
            [("BUY_ME".to_string(), 0.6), ("SELL_ME".to_string(), 0.2)]
                .into_iter()
                .collect();
        let current: HashMap<String, f64> =
            [("BUY_ME".to_string(), 0.3), ("SELL_ME".to_string(), 0.5)]
                .into_iter()
                .collect();
        let result = rebalancer.rebalance(&target, &current, 1_000.0);
        let buy = result.trades.iter().find(|t| t.symbol == "BUY_ME").unwrap();
        let sell = result
            .trades
            .iter()
            .find(|t| t.symbol == "SELL_ME")
            .unwrap();
        assert_eq!(buy.side, "BUY");
        assert_eq!(sell.side, "SELL");
    }
}
